oxd-server/src/test/java/org/gluu/oxd/server/service/RpSyncServiceTest.java

```diff
@@ -0,0 +1,117 @@
+import io.dropwizard.jersey.validation.Validators;
+import io.dropwizard.testing.ResourceHelpers;
+import org.gluu.oxd.server.OxdServerConfiguration;
+import org.gluu.oxd.server.Utils;
+import org.gluu.oxd.server.guice.GuiceModule;
+import org.gluu.oxd.server.persistence.PersistenceService;
+import org.testng.Assert;
+import org.testng.annotations.AfterSuite;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Guice;
+import org.testng.annotations.Test;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Calendar;
+import java.util.Date;
+
+@Guice(modules = GuiceModule.class)
+public class RpSyncServiceTest {
+
+    @Inject
+    ConfigurationService configurationService;
+    @Inject
+    RpService service;
+    @Inject
+    RpSyncService rpSyncService;
+    @Inject
+    PersistenceService persistenceService;
+
+    @BeforeClass
+    public void setUp() throws IOException, ConfigurationException {
+        configurationService.setConfiguration(parseConfiguration(ResourceHelpers.resourceFilePath("oxd-server-jenkins.yml")));
+        persistenceService.create();
+        service.removeAllRps();
+        service.load();
+    }
+
+    @AfterSuite
+    public void tearDownSuite() {
+        persistenceService.destroy();
+    }
+
+    @Test
+    public void testShouldSyncForNullRp() {
+        Assert.assertFalse(rpSyncService.shouldSync(null));
+        service.removeAllRps();
+    }
+
+    @Test
+    public void testShouldSyncForNotNullRp() {
+        Rp rp = new Rp();
+        rp.setSyncClientFromOp(true);
+        //adding date 10 mintutes behind current time and setting as LastSynced
+        rp.setLastSynced(Utils.addTimeToDate(new Date(), -10, Calendar.MINUTE));
+        rp.setSyncClientPeriodInSeconds(5 * 60);
+        Assert.assertTrue(rpSyncService.shouldSync(rp));
+    }
+
+    @Test
+    public void testShouldSyncForSyncAllowed() {
+        Rp rp = new Rp();
+        rp.setSyncClientFromOp(true);
+        //adding date 10 mintutes behind current time and setting as LastSynced
+        rp.setLastSynced(Utils.addTimeToDate(new Date(), -10, Calendar.MINUTE));
+        rp.setSyncClientPeriodInSeconds(5 * 60);
+        Assert.assertTrue(rpSyncService.shouldSync(rp));
+    }
+
+    @Test
+    public void testShouldSyncForSyncNotAllowed() {
+        Rp rp = new Rp();
+        rp.setSyncClientFromOp(false);
+        //adding date 10 mintutes behind current time and setting as LastSynced
+        rp.setLastSynced(Utils.addTimeToDate(new Date(), -10, Calendar.MINUTE));
+        rp.setSyncClientPeriodInSeconds(5 * 60);
+        Assert.assertFalse(rpSyncService.shouldSync(rp));
+    }
+
+    @Test(expectedExceptions = {NullPointerException.class})
+    public void testShouldSyncForNullLastSynced() {
+        Rp rp = new Rp();
+        rp.setSyncClientFromOp(true);
+        rp.setLastSynced(null);
+        rp.setSyncClientPeriodInSeconds(5 * 60);
+        Assert.assertTrue(rpSyncService.shouldSync(rp));
```

Comment:
We should try to write NPE free code. Otherwise it can quickly become spaghetti.

Here if last synced is null then we never synced it yet. (Because if it was synced then it must be not null). It means that if null then `shouldSync` should return true (without NPE).

Reply:
Done